s3/src/main/scala/akka/stream/alpakka/s3/auth/CanonicalRequest.scala

```diff
@@ -43,14 +65,21 @@ private[alpakka] object CanonicalRequest {
    * percent encoded path delimiters back to their decoded counterparts.
    */
   private def preprocessPath(path: Path): String =
-    uriEncode(path.toString).replace(":", "%3A").replace("%2F", "/")
+    uriEncode(removeRedundantAndRelativePath(path.toString))
+      .replace(":", "%3A")
+      .replace("%2F", "/")
+      .replace("%7E", "~")
+      .replace("+", "%20")
 
   def canonicalHeaderString(headers: Seq[HttpHeader]): String = {
     val grouped = headers.groupBy(_.lowercaseName())
-    val combined = grouped.mapValues(_.map(_.value.replaceAll("\\s+", " ").trim).mkString(","))
-    combined.toList.sortBy(_._1).map { case (k, v) => s"$k:$v" }.mkString("\n")
+    val combined = grouped.mapValues(_.map(_.value.replaceAll("\\s+", " ")).mkString(","))
+    fixContentTypeHeaderParameter(combined.toList.sortBy(_._1).map { case (k, v) => s"$k:$v" }.mkString("\n"))
   }
 
+  private def fixContentTypeHeaderParameter(headers: String): String =
+    headers.replace("charset=UTF-8", "charset=utf8")
```

Comment:
This will break signing, as you sign a different request than what akka-http will send.